api/rate/rate.go

```diff
@@ -111,6 +115,10 @@ func (l *Limiter) Acquire(ctx context.Context, path string) error {
 	}
 
 	if sleep > 0 {
+		if deadline, ok := ctx.Deadline(); ok && time.Now().Add(sleep).After(deadline) {
```

Comment:
Instead of recalculating the time, we should check the deadline against the time that we already have. That should be more accurate.